### ISSUE TITLE
client-go/util/workqueue: DelayingQueue does not deduplicate items added with a zero or very small duration

### ISSUE THREAD

**Issue by inteon** [kubernetes/kubernetes#126027]
We encountered some very unexpected behavior while working with the [DelayingQueue](https://github.com/kubernetes/kubernetes/blob/5a99930ab57b2c7837960b21d2c92e958db8c377/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go#L68).
Calling `q.AddAfter(item, 5*time.Hour)` and `q.AddAfter(item, 10*time.Millisecond)` will result in a single queue item being added after 10ms.
However, calling `q.AddAfter(item, 5*time.Hour)` and `q.AddAfter(item, 0*time.Millisecond)` results in two  items added, one after 0ms and one after 5h (instead of just once item at 0ms).

The same happens when using `q.Add(item)` instead of `q.AddAfter(item, 0*time.Millisecond)`.

This behavior can also be triggered when using the RateLimitingQueue in combination with an update event from eg. an informer.

The cause is a short-circuit that is triggered [here](https://github.com/kubernetes/kubernetes/blob/5a99930ab57b2c7837960b21d2c92e958db8c377/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go#L241-L245) or [here](https://github.com/kubernetes/kubernetes/blob/5a99930ab57b2c7837960b21d2c92e958db8c377/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go#L315-L317). Ideally, in both places, there would be a "remove from heap" operation. Similarly, `q.Add(...)` does not trigger a "remove from heap" operation.

The following test code can be used to test this behavior (add this to `delaying_queue_test.go`):
```go
func TestDeduping(t *testing.T) {
	fakeClock := testingclock.NewFakeClock(time.Now())
	q := NewDelayingQueueWithConfig(DelayingQueueConfig{Clock: fakeClock})

	first := "foo"
	initialDelay := 1 * time.Millisecond // 0ms makes the test fail here.

	q.AddAfter(first, initialDelay+0*time.Millisecond)
	if err := waitForWaitingQueueToFill(q); err != nil {
		t.Fatalf("unexpected err: %v", err)
	}
	q.AddAfter(first, initialDelay+50*time.Millisecond)
	if err := waitForWaitingQueueToFill(q); err != nil {
		t.Fatalf("unexpected err: %v", err)
	}

	// step past the first block, we should receive now
	fakeClock.Step(10 * time.Millisecond)
	if err := waitForWaitingQueueToFill(q); err != nil {
		t.Fatalf("unexpected err: %v", err)
	}
	if q.Len() != 1 {
		t.Error("should have added")
	}
	item, _ := q.Get()
	q.Done(item)

	// step past the second add
	fakeClock.Step(50 * time.Millisecond)
	if err := waitForWaitingQueueToFill(q); err != nil {
		t.Fatalf("unexpected err: %v", err)
	}
	if q.Len() != 0 {
		t.Errorf("should not have added")
	}
}
```

**Comment by inteon** [comment]
> If you add something immediately and something after X secs, why should dedup the immediate?

It should remove the scheduled item, since we already processed it now.

**Comment by aojea** [comment]
I see, so is this because you want to enqueue something, let's say after X seconds, but you may have another event that indicates you can process it immediately and you don't have to way, so you do want to run it now and stop to run it again after X seconds?

**Comment by inteon** [comment]
> I see, so is this because you want to enqueue something, let's say after X seconds, but you may have another event that indicates you can process it immediately and you don't have to way, so you do want to run it now and stop to run it again after X seconds?

Indeed, for example when there is an error that causes an exponential backoff, but then the resource is updated and the resource is reconciled immediately. In that case, I expect the scheduled enqueue to be canceled.

**Comment by aojea** [comment]
Just curious, I'm AFK so asking for experimentation, what happen if you add the second event with a delay of 1 millisecond?

**Comment by inteon** [comment]
> Just curious, I'm AFK so asking for experimentation, what happen if you add the second event with a delay of 1 millisecond?

In that case, the 5 hour delay will be replaced by the 1ms delay.
The existing scheduled item is not removed when the delay is <=0 or so small that the delay passed before this check: https://github.com/kubernetes/kubernetes/blob/5a99930ab57b2c7837960b21d2c92e958db8c377/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go#L315.

**Comment by aojea** [comment]
It seems we have a workaround then

IMHO changing the current semantics of `Add()` to cancel the waiting item will be a breaking behavior, as Add() is explicitly shortcutting the queue today
